src/pdftoraw.c

```diff
@@ -0,0 +1,209 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <unistd.h>
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <getopt.h>
+#include <gio/gio.h>
+#include <poppler.h>
+#include <gdk/gdk.h>
+
+
+// 1 inch = 72 points
+#define PTS 72.0
+
+// Resolution for the printer, 150 my choice
+#define DPI 150.0
+
+static gchar *filename = "";
+static gboolean opt_test = FALSE;
+static gboolean opt_pages = FALSE;
+static int opt_raw = -1;
+static int opt_width = -1;
+static int opt_height = -1;
+
+static GOptionEntry entries[] = {
+  { "file", 'f', 0, G_OPTION_ARG_STRING, &filename, "PDF file path", NULL },
+  { "test", 't', 0, G_OPTION_ARG_NONE, &opt_test, "Test if the file is a PDF file", NULL },
+  { "pages", 'p', 0, G_OPTION_ARG_NONE, &opt_pages, "Returns the number of pages in the PDF file", NULL },
+  { "width", 'w', 0, G_OPTION_ARG_INT, &opt_width, "Give the size of the page provided in arguments", NULL},
+  { "height", 'W', 0, G_OPTION_ARG_INT, &opt_height, "Gives the height of the page provided in arguments", NULL},
+  { "raw", 'r', 0, G_OPTION_ARG_INT, &opt_raw, "Retrieves data in pixels from the page provided as arguments", NULL},
+  { NULL }
+};
+
+static void
+print_usage (void)
+{
+    printf("Usage:\tpdftoraw  --filename=<filename>  --raw=<num-page>\n");
+    printf("\tpdftoraw  --filename=<filename>  --test\n");
+    printf("\tpdftoraw  --filename=<filename>  --pages\n");
+    printf("\tpdftoraw  --filename=<filename>  --width=<num-page>\n");
+    printf("\tpdftoraw  --filename=<filename>  --height=<num-page>\n");
+    printf("\tpdftoraw  --help\n");
+}
+
+static int
+pdf_number_pages_get (PopplerDocument *doc)
+{
+    return poppler_document_get_n_pages(doc);
+}
+
+static void
+pdf_page_raw_get (PopplerDocument *doc,
+                 int page_nr)
+{
+    cairo_surface_t *s = NULL;
+    cairo_t *cr = NULL;
+    unsigned char *data = NULL;
+    PopplerPage *page = NULL;
+    g_autoptr(GdkPixbuf) pix = NULL;
+    double width = 0;
+    double height = 0;
+    int w = 0;
+    int h = 0;
+
+    page = poppler_document_get_page(doc, page_nr);
+    if (page == NULL)
+       return;
+    poppler_page_get_size(page, &width, &height);
+    width = DPI * width / PTS;
+    height = DPI * height / PTS;
+    w = (int)ceil(width);
+    h = (int)ceil(height);
+    s = cairo_image_surface_create(CAIRO_FORMAT_ARGB32, w, h);
+    if (s == NULL)
+       return;
```

Comment:
In these exceptional cases can you add a `g_debug ("Foo failed");`.

Reply:
Fixed by 9cb421f